File: algorithms/cpp/ConstructStringFromBinaryTree/construct_string_from_binary_tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 public:
  string tree2str(TreeNode* t) {
    if (0 == t) {
      return string();
    }
    string left_str = tree2str(t->left);
    string right_str = tree2str(t->right);
    ostringstream string_stream;
    if (0 < right_str.size()) {
      string_stream << t->val << "(" << left_str << ")"
                    << "(" << right_str << ")";
    } else if (0 == left_str.size()) {
      string_stream << t->val;
    } else {
      string_stream << t->val << "(" << left_str << ")";
    }
    return string_stream.str();
  }
};

File: algorithms/py/ConstructStringFromBinaryTree/construct_string_from_binary_tree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if t is None:
            return ""
        
        left_str = self.tree2str(t.left)
        right_str = self.tree2str(t.right)
        
        if len(right_str) > 0:
            return "%s(%s)(%s)" % (str(t.val), left_str, right_str)
        elif len(left_str) == 0:
            return str(t.val)
        else:
            return "%s(%s)" % (str(t.val), left_str)
